src/internal/storage/kv/filesystem.go

```diff
@@ -0,0 +1,206 @@
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+	"go.uber.org/zap"
+	"golang.org/x/sync/semaphore"
+)
+
+var _ Store = &FSStore{}
+
+type FSStore struct {
+	dir          string
+	maxKeySize   int
+	maxValueSize int
+
+	// initDone is true after the store has been initialized
+	initDone atomic.Bool
+	initSem  *semaphore.Weighted
+}
+
+func NewFSStore(dir string, maxKeySize, maxValueSize int) *FSStore {
+	return &FSStore{
+		dir:          dir,
+		maxKeySize:   maxKeySize,
+		maxValueSize: maxValueSize,
+		initSem:      semaphore.NewWeighted(1),
+	}
+}
+
+func (s *FSStore) Put(ctx context.Context, key, value []byte) (retErr error) {
+	if len(key) > s.maxKeySize {
+		return errors.Errorf("max key size %d exceeded. len(key)=%d", s.maxKeySize, len(key))
+	}
+	if len(value) > s.maxValueSize {
+		return errors.Errorf("max value size %d exceeded. len(value)=%d", s.maxKeySize, len(value))
+	}
+	if err := s.ensureInit(ctx); err != nil {
+		return err
+	}
+	staging := s.stagingPathFor(key)
+	final := s.finalPathFor(key)
+	defer s.cleanupFile(ctx, &retErr, staging)
+	if err := os.WriteFile(staging, value, 0o644); err != nil {
+		return s.transformError(err, key)
+	}
+	return errors.EnsureStack(os.Rename(staging, final))
+}
+
+func (s *FSStore) Get(ctx context.Context, key, buf []byte) (_ int, retErr error) {
+	f, err := os.Open(s.finalPathFor(key))
+	if err != nil {
+		return 0, s.transformError(err, key)
+	}
+	defer s.closeFile(ctx, &retErr, f)
+	return miscutil.ReadInto(buf, f)
+}
+
+func (s *FSStore) Exists(ctx context.Context, key []byte) (bool, error) {
+	_, err := os.Stat(s.finalPathFor(key))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, s.transformError(err, key)
+	}
+	return true, nil
+}
+
+func (s *FSStore) Delete(ctx context.Context, key []byte) error {
+	if err := s.ensureInit(ctx); err != nil {
+		return err
+	}
+	err := os.Remove(s.finalPathFor(key))
+	if os.IsNotExist(err) {
+		err = nil
+	}
+	return err
+}
+
+func (s *FSStore) stagingPathFor(k []byte) string {
+	return filepath.Join(s.dir, "staging", uuid.NewWithoutDashes())
+}
+
+func (s *FSStore) finalPathFor(k []byte) string {
+	return filepath.Join(s.dir, "objects", hex.EncodeToString(k))
+}
+
+func (s *FSStore) NewKeyIterator(span Span) stream.Iterator[[]byte] {
+	return &fsIterator{s: s, span: span}
+}
+
+type fsIterator struct {
+	s    *FSStore
+	span Span
+	keys [][]byte
+	pos  int
+}
+
+func (it *fsIterator) Next(ctx context.Context, dst *[]byte) error {
+	if it.keys == nil {
+		dirEnts, err := os.ReadDir(filepath.Join(it.s.dir, "objects"))
+		if err != nil {
+			return errors.EnsureStack(err)
+		}
+		var keys [][]byte
+		for i := range dirEnts {
+			key, err := hex.DecodeString(dirEnts[i].Name())
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
+			keys = append(keys, key)
+		}
+		it.keys = keys
+	}
+	for it.pos < len(it.keys) {
+		k := it.keys[it.pos]
+		if it.span.Contains(k) {
+			*dst = append((*dst)[:0], k...)
```

Comment:
This looks like a complicated way of copying the contents of `k` into `*dst` by value. It may be more readable to do something like the following:
``` Go
*dst = make([]byte, len(key))
copy(*dst, key)
```

Reply:
^That will allocate a new buffer for dst, rather than reusing the existing buffer.